backend/src/main/kotlin/app/ehrenamtskarte/backend/verification/CanonicalJson.kt

```diff
@@ -0,0 +1,109 @@
+                    // older protos without the field:
+                    // If the repeated field is empty (if there are no elements "in the array"), do not emit anything in the JSON.
+                    // If there is at least one field, emit a JSON Array by mapping each element to its JSON equivalent (see below).
+                    throw Error("Repeated fields are currently not supported.")
+                }
+            }
+        }
+
+        private fun GeneratedMessageV3.assertMessageIsValid() {
+            assertOnlyKnownFields()
+            assertOnlyOptionalFields()
+            assertNoRepeatedFields()
+        }
+
+        /**
+         * Returns a canonical JSON object from the given protobuf message.
+         * For every present field in the message, we add an entry to the resulting object where the key is the tagNumber of the
+         * field, and the value is the appropriately encoded JSON value.
+         * Note, that we encode integers as strings because JSON-Number does not allow the full range of uint64 integers.
+         */
+        fun messageToMap(message: GeneratedMessageV3): Map<String, Any> {
+            message.assertMessageIsValid()
+
+            return message.allFields.entries.filter { it.value != null }.associate {
+                return@associate when (it.key.type) {
+                    Type.STRING, Type.BOOL, Type.SINT32, Type.INT32, Type.UINT32, Type.UINT64 -> Pair(
+                        it.key.number.toString(),
+                        it.value.toString()
+                    )
+
+                    Type.ENUM -> {
+                        val enumValue = it.value
+                        if (enumValue !is Descriptors.EnumValueDescriptor) {
+                            throw Error("Field ${it.key.name} is not an instance of Enum despite its kind is 'enum'.")
+                        }
+                        Pair(it.key.number.toString(), enumValue.number.toString())
+                    }
+
+                    Type.MESSAGE -> {
+                        val subMessage = it.value
+                        if (subMessage !is GeneratedMessageV3) {
+                            throw Error("Field ${it.key.name} is not an instance of Message despite its kind is 'message'.")
+                        }
+                        Pair(it.key.number.toString(), messageToMap(subMessage))
+                    }
+
+                    else -> throw Error("Field ${it.key.name} has an unsupported type of ${it.key.type}")
+                }
+            }
+        }
+
+        fun serializeToString(message: GeneratedMessageV3) = serializeToString(messageToMap(message))
+        private fun Int.isSafeInteger() = this > lowestSafeInt && this < highestSafeInt
+
+        fun serializeToString(o: Any?): String {
+            return when (o) {
+                null -> "null"
+                is String -> "\"${o}\""
```

Comment:
This is not sufficient! You need to escape some characters. It would be better to use some existing JSON Encoder for Strings, numbers, etc.

Reply:
Maybe also add a test case that includes a string containing `"` or containing a new line char.